sop-core/sop-core.cabal

```diff
@@ -1,5 +1,5 @@
 name:                sop-core
-version:             0.5.0.1
+version:             0.5.1.0
```

Comment:
FWIW, this is definitely not a non-breaking change.

Reply:
OTOH, `staged-streams` compiles fine. (Plenty of warnings about non-exhautive pattern matches on `NS`, but those are not fatal).

Reply:
Yeah, my _hope_ was that with the pattern synonym it would be backwards compatible; and those warnings should even be absent from 8.10.

Reply:
The data structure is now stricter. This is contrived example, but still, the behavior changes observably:

Currently,

```haskell
hd $ I 'x' :* (error "ay, caramba!" :: NP I '[])
I 'x'
```

With this patch

```haskell
*Data.SOP Data.SOP.NP> hd $ I 'x' :* (error "ay, caramba!" :: NP I '[])
I *** Exception: ay, caramba!
CallStack (from HasCallStack):
  error, called at <interactive>:21:16 in interactive:Ghci1
```

Reply:
Ah, fair enough. Hadn't thought terribly hard about strictness just yet :)

Reply:
Is there a good reason why someone might rely on the construction of the NP being lazy? the idea of writing code which might only look at the first N elements of a NP seems pretty odd to me.

Reply:
FWIW, I think that `NP` and `NS` *should* be spine-strict, and not making them so from the beginning was a mistake. Before `sop-core`, I'd have said there's basically no reason not to make the change. Now, it's theoretically possible there are uses that aren't generic programming related which might have other requirements. But it may be worth just doing it.